### PR TITLE
[stdlib] Implement the new _modify accessor in Dictionary.subscript

### PR DESCRIPTION
**Pull request by lorentey**
This enables in-place mutations of the key-based subscript.
These get rid of one of two full hash table lookup operations (including hashing the key) relative to the getter+setter approach. They can also eliminate COW copies in code like this:

```swift
dictionary[foo]?.append(bar)
```

(Admittedly this isn’t a very useful example.)

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci smoke test

**Comment by lorentey** [comment]
Tests, motivating examples are not yet done.

**Comment by slavapestov** [comment]
General question for @rjmccall - are we going to use this with arrays too, or are we going to keep addressors around as a language feature and synthesize a ‘modify’ out of them?

**Comment by rjmccall** [comment]
Addressors can't completely disappear because we have to bootstrap unsafe pointers somehow.  But yes, #18840 can synthesize a modify from a mutable addressor.

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark staging

**Comment by airspeedswift** [comment]
Huh. If we really were avoiding double-hashing, I'd have expected more signal from the benchmarks. Surely we must be doing lots of dictionary subscript assigns in them.

**Comment by lorentey** [comment]
We have benchmarks setting values, but simple subscript assigns (`dict[foo] = bar`) already hash just once - they don’t benefit from the new accessor. To demonstrate an improvement, we need to find an algorithm that performs in-place mutations on Optionals.

File: stdlib/public/core/Dictionary.swift [hunk 1]
```diff
@@ -785,6 +785,60 @@ extension Dictionary {
...
+      let isUnique = _variant.isUniquelyReferenced()
+      var idealBucket = _variant.asNative._bucket(key)
+      var (pos, found) = _variant.asNative._find(key, startBucket: idealBucket)
+
+      // Prepare storage.
+      // If `key` isn't in the dictionary yet, assume that this access will end
+      // up inserting it. Otherwise this can only be a removal or an in-place
+      // mutation. (If we guess wrong, we might needlessly rehash; that's fine.)
+      let (_, rehashed) = _variant.ensureUniqueNative(
+        withCapacity: self.capacity + (found ? 0 : 1),
```
Comment (lorentey) [inline stdlib/public/core/Dictionary.swift]: Good thing benchmarks don’t exercise this yet — this should use self.count, not capacity. (As it stands, this forces a full 2x storage reallocation & rehash on every insertion.) 😳

**Comment by airspeedswift** [comment]
I guess the trouble is that nearly all the useful in-place mutations happen via the `subscript(_:default:)` version. `dictionary[foo]?.append(bar)` isn't useful but `dictionary[foo, default: []).append(bar)` is.

**Comment by jrose-apple** [comment]
You should be able to use basically the same logic for `subscript(_:default:)`, no? By the time you get to a `modify` accessor, you know you want to do the writeback.

**Comment by rjmccall** [comment]
`subscript(_:default:)` is already implemented with a mutable addressor.  It could switch to a `read` instead of a getter, but that won't show any benefits until we optimize the SILGen of `read` a little.

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
Trying again now #19085 is merged.

@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please smoke test

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
Bout to merge this, just testing it one more time since it's been a couple of weeks

**Comment by lorentey** [comment]
@swift-ci smoke test and merge

**Pull request merged**